impl/memory/mutation_logs.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package memory supplies fake in-memory implementations for testing purposes.
+package memory
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/google/keytransparency/core/mutator"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+)
+
+// NewMutationLog creates a new fake MutationLog
+func NewMutationLog() MutationLog {
+	return make(MutationLog)
+}
+
+// MutationLog is a fake implementation of keyserver.
+type batch struct {
+	time time.Time
+	msgs []*mutator.LogMessage
+}
+
+// MutationLog is a fake, in-memory implementation of keyserver.MutationLogs
+type MutationLog map[int64][]batch // Map of logID to Slice of LogMessages
+
+// AddLogs to the mutation database.
+func (m MutationLog) AddLogs(ctx context.Context, directoryID string, logIDs ...int64) error {
+	for _, logID := range logIDs {
+		m[logID] = make([]batch, 0)
+	}
+	return nil
+}
+
+// AddLogs to the mutation database.
+func (m MutationLog) ListLogs(ctx context.Context, directoryID string, writable bool) ([]int64, error) {
+	logIDs := []int64{}
+	for logID := range m {
+		logIDs = append(logIDs, logID)
+	}
+	sort.Slice(logIDs, func(a, b int) bool { return logIDs[a] < logIDs[b] })
+	return logIDs, nil
+}
+
+// Send a batch of mutations to a random log.
+func (m MutationLog) Send(ctx context.Context, dirID string, mutation ...*pb.EntryUpdate) (int64, time.Time, error) {
+	// Select a random logID
+	var logID int64
+	for i := range m {
+		logID = i
+		break
+	}
+	ts := time.Now()
+	// Remove the PII contained in EntryUpdate. Only save the merkle tree bits.
+	entries := make([]*pb.SignedEntry, 0, len(mutation))
+	for _, i := range mutation {
+		entries = append(entries, i.Mutation)
+	}
+	m.SendAt(logID, ts, entries)
+	return logID, ts, nil
+}
+
+// SendAt sends a batch of mutations with a given timestamp.
+func (m MutationLog) SendAt(logID int64, ts time.Time, entries []*pb.SignedEntry) {
+	logShard := m[logID]
+	if len(logShard) > 0 && logShard[len(logShard)-1].time.After(ts) {
+		panic(fmt.Sprintf("inserting mutation entry %v out of order", ts))
```

Comment:
This function is only called in test setup so I was trying to avoid extra error checking. 
I'll make this more clearly at testing function by passing t.

Reply:
Note: this solution requires a bit of a hack in `Send` let me know if that's not acceptable.